crates/feldera-storage/src/file/writer.rs

```diff
@@ -0,0 +1,1031 @@
+}
+
+impl ColumnWriter {
+    fn new(parameters: &Rc<Parameters>, column_index: usize) -> Self {
+        ColumnWriter {
+            parameters: parameters.clone(),
+            column_index,
+            rows: 0..0,
+            data_block: DataBlockBuilder::new(parameters),
+            index_blocks: Vec::new(),
+        }
+    }
+
+    fn take_rows(&mut self) -> Range<u64> {
+        let end = self.rows.end;
+        replace(&mut self.rows, end..end)
+    }
+
+    fn finish<W, K, A>(&mut self, writer: &mut W) -> IoResult<FileTrailerColumn>
+    where
+        W: Write + Seek,
+        K: Rkyv,
+        A: Rkyv,
+    {
+        // Flush data.
+        if !self.data_block.is_empty() {
+            let data_block = self.data_block.take().build();
+            self.write_data_block::<W, K, A>(writer, data_block)?;
+        }
+
+        // Flush index.
+        let mut retval = None;
+        for level in 0..self.index_blocks.len() {
+            if !self.index_blocks[level].is_empty() {
+                let index_block = self.index_blocks[level].take().build();
+                retval = Some(self.write_index_block::<W, K>(writer, index_block, level)?);
+            }
+        }
+        if let Some((location, n_rows)) = retval {
+            Ok(FileTrailerColumn {
+                index_offset: location.offset,
+                index_size: location.size as u32,
+                n_rows,
+            })
+        } else {
+            Ok(FileTrailerColumn {
+                index_offset: 0,
+                index_size: 0,
+                n_rows: 0,
+            })
+        }
+    }
+
+    fn get_index_block(&mut self, level: usize) -> &mut IndexBlockBuilder {
+        if level >= self.index_blocks.len() {
+            debug_assert_eq!(level, self.index_blocks.len());
+            self.index_blocks.push(IndexBlockBuilder::new(
+                &self.parameters,
+                self.column_index,
+                if level == 0 {
+                    NodeType::Data
+                } else {
+                    NodeType::Index
+                },
+            ));
+        }
+        &mut self.index_blocks[level]
+    }
+
+    fn write_data_block<W, K, A>(
+        &mut self,
+        writer: &mut W,
+        mut data_block: DataBlock,
+    ) -> IoResult<()>
+    where
+        W: Write + Seek,
+        K: Rkyv,
+        A: Rkyv,
+    {
+        let location = write_block(writer, &mut data_block.raw)?;
+
+        let serialize_minmax = |dst: &mut AlignedVec| {
+            let min_offset = rkyv_copy_key::<K, A>(dst, &data_block.raw, data_block.min_offset);
+            let max_offset = rkyv_copy_key::<K, A>(dst, &data_block.raw, data_block.max_offset);
+            (min_offset, max_offset)
+        };
+        if let Some(index_block) = self.get_index_block(0).add_entry(
+            location,
+            serialize_minmax,
+            data_block.n_values as u64,
+        ) {
+            self.write_index_block::<W, K>(writer, index_block, 0)
```

Comment:
what's the point of the map?

Reply:
It discards the return value from `write_index_block`, converting `IoResult<(BlockLocation, u64)>` into `IoResult<()>` for this function to return.

I applied the following, which is probably clearer:

```patch
diff --git a/crates/feldera-storage/src/file/writer.rs b/crates/feldera-storage/src/file/writer.rs
index b11a82a5..e89ff7b8 100644
--- a/crates/feldera-storage/src/file/writer.rs
+++ b/crates/feldera-storage/src/file/writer.rs
@@ -260,11 +290,9 @@ impl ColumnWriter {
             serialize_minmax,
             data_block.n_values as u64,
         ) {
-            self.write_index_block::<W, K>(writer, index_block, 0)
-                .map(|_| ())
-        } else {
-            Ok(())
+            self.write_index_block::<W, K>(writer, index_block, 0)?;
         }
+        Ok(())
     }
 
     fn write_index_block<W, K>(
```